Python3 conversion

aaronchantrill opened the pull request:
## Description
Rewrite naomi-setup.sh to install and link to python3. Also
fixed an issue at the end of the setup where it writes the final
./Naomi file where it was not putting a literal $@ at the end
of the line, so not properly passing command line arguments through.

Added a line to .gitignore to ignore the ./Naomi file created by
naomi-setup.sh

Set crunch-bang line in Naomi.py to use python3 instead of python2.

Changed ```./naomi/__main__.py``` to set *_MIC globals from application
rather than just importing them, because the relative import did
not work. There are a few ways to fix that, but this seemed easiest
for right now.

./naomi/i18n.py, ./naomi/mic.py, ./plugins/tts/flite-tts/flite.py,
./plugins/stt/pocketsphinx-stt/g2p.py,
./plugins/stt/pocketsphinx-stt/sphinxplugin.py,
./plugins/stt/pocketsphinx-stt/sphinxvocab.py
Changed some encoding, fixing new errors around str to byte
conversions.

./plugins/stt/pocketsphinx-stt/sphinxplugin.py
Changed the import from

```import pocketsphinx```

to

```from pocketsphinx import pocketsphinx```

also added ```from sphinxbase.sphinxbase import *``` because otherwise
the line ```config = pocketsphinx.Decoder.default_config()```
did not work. I'm not sure why that is necessary. I am currently
running the git version of pocketsphinx, so I don't know if there
are different problems around earlier versions, or why you have
to import pocketsphinx.pocketsphinx now instead of just
pocketsphinx.

./plugins/speechhandler/joke/data/en-US.txt
restored file from Jasper original file. I'm not sure why this was
removed. I did not restore the german joke file. I don't think
there has ever been a french joke file.

./plugins/speechhandler/weather/weather.py
discovered that weather is reading the location from profile.yml at
```
weather:
  location:
```
also converting the location to woeid, which can be stored at
```
weather:
  woeid:
```
so I added an info line stating what the woeid is so it can
be put in the profile.yml file and the conversion step can be
skipped.

./populate.py
Changed the location of the weather plugin "location" variable
from
```location:```
to
```
weather:
  location:
```

My copy of Naomi is now running on python3 using pocketsphinx
and flite. Overall, this was a lot easier than I thought it was
going to be, and mostly just involved running Naomi.py and
waiting for it to crash with an error about needing a byte
object and finding a string object (add b before the quotes
or add ".encode('utf-8')") or needing a string and finding a
byte object (add .decode('utf-8') to the end).

## Related Issue
[Migrate to Python 3 soon](https://github.com/NaomiProject/Naomi/issues/90)

## Motivation and Context
We need to migrate to python 3 because python 2 is deprecated. Better sooner
than later.

## How Has This Been Tested?
Note: a lot of the unit tests are failing right now. I will work on those later.

I tested using the git version of pocketsphinx from the pocketsphinx install
instructions: https://projectnaomi.com/docs/plugins/pocketsphinx-install.html
and flite installed from apt, and using alsa as the engine. I used the 
./naomi-setup.sh script to create the environment. I am testing on a copy of
Raspbian that I also used for testing the german language files, so I still
need to test on a clean copy of Raspbian.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
Changed from python2 to python3 to start the process of conversion, but I'm sure this change
breaks a lot of things that I have not tested yet. A lot of the unit tests are currently failing. I will
take a look at them this afternoon.

## Checklist:
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
I have updated the documentation.

The python3 version of unittest works slightly differently from the python2 version of unittest. The python 3 version appears to take it upon itself to load every module under the project directory as a test. Some of these tests are failing due to not having python modules or linux libraries installed. I have loaded every required module and library onto a Raspberry Pi and the only tests that still failed were osx-tts (failed because not running on osx) and mpdcontrol (because of a python3 compatibility issue in the mpd module).

aaronchantrill committed:
Starting Python3 conversion

Rewrite naomi-setup.sh to install and link to python3. Also
fixed an issue at the end of the setup where it writes the final
./Naomi file where it was not putting a literal $@ at the end
of the line, so not properly passing command line arguments through.

Added a line to .gitignore to ignore the ./Naomi file created by
naomi-setup.sh

Set crunch-bang line in Naomi.py to use python3 instead of python2.

Changed ./naomi/__main__.py to set *_MIC globals from application
rather than just importing them, because the relative import did
not work. There are a few ways to fix that, but this seemed easiest
for right now.

./naomi/i18n.py, ./naomi/mic.py, ./plugins/tts/flite-tts/flite.py,
./plugins/stt/pocketsphinx-stt/g2p.py,
./plugins/stt/pocketsphinx-stt/sphinxplugin.py,
./plugins/stt/pocketsphinx-stt/sphinxvocab.py
Changed some encoding, fixing new errors around str to byte
conversions.

./plugins/stt/pocketsphinx-stt/sphinxplugin.py
Changed the import from

import pocketsphinx

to

from pocketsphinx import pocketsphinx

also added "from sphinxbase.sphinxbase import *" because otherwise
the line "config = pocketsphinx.Decoder.default_config()"
did not work. I'm not sure why that is necessary. I am currently
running the git version of pocketsphinx, so I don't know if there
are different problems around earlier versions, or why you have
to import pocketsphinx.pocketsphinx now instead of just
pocketsphinx.

./plugins/speechhandler/joke/data/en-US.txt
restored file from Jasper original file. I'm not sure why this was
removed. I did not restore the german joke file. I don't think
there has ever been a french joke file.

./plugins/speechhandler/weather/weather.py
discovered that weather is reading the location from profile.yml at
weather:
  location:
also converting the location to woeid, which can be stored at
weather:
  woeid:
so I added an info line stating what the woeid is so it can
be put in the profile.yml file and the conversion step can be
skipped.

./populate.py
Changed the location of the weather plugin "location" variable
from

location:

to

weather:
  location:

My copy of Naomi is now running on python3 using pocketsphinx
and flite. Overall, this was a lot easier than I thought it was
going to be, and mostly just involved running Naomi.py and
waiting for it to crash with an error about needing a byte
object and finding a string object (add b before the quotes
or add ".encode('utf-8')") or needing a string and finding a
byte object (add .decode('utf-8') to the end).

aaronchantrill committed:
Fixed Naomi3 to pass unit tests

After converting Naomi from python2 to python3, a lot of the
unit tests started failing. More than were even performed under
python2. I went through and got all but five to pass so far.
Those appear to be failing because I do not have them configured.
I think I can configure unittest to skip any options that have
not been configured, but that puts us in a weird situation where
if someone changes something that affects an option they don't
have configured, the test that would have showed the problem
gets skipped.

Here are the modules that failed for me and the reasons:
espeak-tts : espeak executable not found
festival-tts : executables "text2wav" and/or "festival" not found
osx-tts : Invalid platform (running on Raspbian)
julius-stt : Can't find julius executable
snowboy-stt : _snowboydetect.so: cannot open shared object file

I believe that the snowboy issue is because I am testing on x86
Raspbian rather than on the Raspberry Pi, and _snowboydetect.so
is precompiled for ARM.

I modified the following files to fix where module names have
changed between python2 and python3:
 naomi/app_utils.py
 plugins/speechhandler/hackernews/hackernews.py
 plugins/speechhandler/news/news.py
 plugins/speechhandler/unclear/unclear.py
 plugins/speechhandler/weather/weather.py
 plugins/stt/google-stt/google.py
 plugins/tts/mary-tts/marytts.py

I discovered that the pocketsphinx tests required that the fst
and acoustic models be in specific directories, so I changed
the tests to use the user's settings instead.
I had to modify the following files:
 naomi/diagnose.py
 naomi/testutils.py
 plugins/stt/pocketsphinx-stt/tests/test_g2p.py
 plugins/stt/pocketsphinx-stt/g2p.py

In pocketsphinx, I also had to force Naomi to create a temp
vocabulary file even when only converting a single word
because phonetisaurus-g2pfst does not feature a one word
argument anymore. This appears to only be a situation that
occurs in testing.

I also fixed an issue where I had stopped forcing everything
to uppercase for the first pass at the g2p converter.

Finally, I fixed an issue in test_g2p.py where the "communicate"
function was returning everything in the format used by
phonetisaurus-g2p, which was failing if the user has configured
phonetisaurus-g2pfst.

I also changed the test wakeword from JASPER to NAOMI and
added a small .wav file to test against.

Finally, in test/test_brain.py I had to change a few strings from
text to unicode.

aaronchantrill committed:
Dumb changes for Codacy

Fixed a bunch of minor Codacy complaints to keep it from flagging
the pull request. Spaces at the ends of lines, variables created
then not used (because I wanted to use a try catch to make sure
the variable being assigned from existed), etc.

aaronchantrill committed:
Resolving conflicts

Resolving two conflicts between current files and previous
hotfix.

aaronchantrill committed:
Fixes for populate

populate.py is a complex program with more text than most
other parts of Naomi. I had to do a lot of switching bytes
to strings and visa versa.

aaronchantrill committed:
Add libncurses5-dev to apt install line

We need libncurses5-dev installed, otherwise the built python
won't be able to build in support for curses which is what we
need for populate.py

aaronchantrill committed:
Remove git and nano from python modules

Don't install git (sometimes causes problems when distros have
their own compilation quirks) or nano (nobody is sure why this
was ever a requirement).

aaronchantrill committed:
Switch python back to python 3.5 in naomi-setup.sh

Switch python back to python 3.5 in naomi-setup.sh
also explicitely set shell=False in populate.py to
see if that works better for Codacy.

aaronchantrill committed:
Merge remote-tracking branch 'refs/remotes/origin/python3' into python3

aaronchantrill committed:
Change flite command to a variable for Codacy

Changing the ["flite","-lv"] command being passed to subprocess
into a variable to see if that makes Codacy happier.

aaronchantrill committed:
Incorporating changes from PR 137

Realized that I was missing some changes from the PR 137 hotfix
especially in naomi/populate.py.

Reincorporated.

aaronchantrill committed:
Resolving differences

Resolving differences between this populate.py and the
current populate.py

aaronchantrill committed:
More codacy changes

Codacy flagged the "shell=False" line again. Trying to change
this to a variable and see if that helps

aaronchantrill committed:
Codacy

Codacy is now flagging things as errors that it passed last
night.

aaronchantrill committed:
Codacy

Still trying to clear the codacy flag.

aaronchantrill committed:
Codacy

Still just juggling things to see how it affects codacy.

aaronchantrill committed:
Codacy

Aarrghh! Why is this happening!

aaronchantrill committed:
Codacy

Still trying.

aaronchantrill committed:
Codacy and subprocess

Still trying to figure out how to get codacy to accept
subprocess.

aaronchantrill committed:
Moved subprocess to own file

Created a file just to wrap subprocess. If I can get this to
work then I'll never need to mess with subprocess and codacy
again.

aaronchantrill committed:
Codacy

Still playing with codacy

aaronchantrill committed:
Codacy and subprocess

Codacy and subprocess

aaronchantrill committed:
Simplifying run_command

No need to create a variable in run_command.

aaronchantrill committed:
Codacy and subprocess

Honestly running out of things to try here.

aaronchantrill committed:
Grasping for straws here

Wow, still can't get Codacy to not flag this.

redragonx commented on naomi/diagnose.py:
```diff
@@ -102,3 +103,40 @@ def get_git_revision():
+# The following were added to be able to use the user's pocketsphinx
+# settings for testing instead of requiring that a model be included
+# with the Naomi source just for testing
+def check_option_in_profile(path):
```
While this is minor, these two methods are very similar in function. I believe in DRY code, I wonder if there is a better way of combing the two? E.G. Pass in an arg to select what you want for the key error part. A boolean would do. 

```
get_profile_val_option(path, True) ...
....
if True then its an option, else its profile option check.
```

aaronchantrill replied:
I agree with DRY and especially preventing data redundancy, but I also believe in usability. I think combining both functions into a single function is going to be a headscratcher for someone coming along afterwards and trying to figure out how to use it. It is true that the diagnose.get_profile_val_option is very similar to the populate.get_profile_var function, so I am looking into pulling the profile walking functions into a separate module. To combine both functions into one, and keep the usability, I would create three functions, two of which call the third that does the heavy lifting. This, however, circumvents what I consider the primary goal of DRY which is reducing the number of unit tests that need to be written to make sure everything works. Now instead of reducing the number of tests from 2 to 1, you still need at least two tests to make sure everything is functioning.

redragonx replied:
Sounds good. Thank you for the explanation.

redragonx commented on naomi-setup.sh:
```diff
@@ -15,26 +15,26 @@ echo "Updating apt, preparing to install libssl-dev, gettext, portaudio19-dev an
-wget https://www.python.org/ftp/python/2.7.13/Python-2.7.13.tgz
-tar xvzf Python-2.7.13.tgz
-cd Python-2.7.13
+wget https://www.python.org/ftp/python/3.5.3/Python-3.5.3.tgz
```
As discussed in Discord. We will create a new PR and issue to add GPG file verification.

aaronchantrill replied:
I tried the GPG verification, and this is what I got:

```
$ gpg --verify Python-3.5.3.tgz.asc gpg: assuming signed data in 'Python-3.5.3.tgz'
gpg: Signature made Tue 17 Jan 2017 03:07:48 AM EST
gpg:                using RSA key 3A5CA953F73C700D
gpg: Good signature from "Larry Hastings <larry@hastings.org>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 97FC 712E 4C02 4BBE A48A  61ED 3A5C A953 F73C 700D
```
I'm not sure if that is an acceptable response or not, not being that familiar with GPG verification of downloaded blobs.

redragonx replied:
This should be an acceptable response, I think there's a way to make the key trusted locally. I didn't do it in my script.

redragonx commented on plugins/stt/pocketsphinx-stt/sphinxvocab.py:
```diff
@@ -10,7 +10,7 @@
 
 def delete_temp_file(file_to_delete):
-    if False:
+    if True:
```
This if statement can be deleted. Appears to be useless atm.

aaronchantrill replied:
You are right, it is only a placeholder. It is helpful for me to sometimes turn off the automatic cleanup of temporary files. Then I can follow the logic through them and figure out why the next step fails exactly. That was the whole purpose of the "delete_temp_file" function, actually. I accidentally left it set to False in an earlier pull request but nobody caught it.

What I would like to do with this is connect it to a command line variable and possibly also a profile setting, so a developer can turn on and off deleting temp files. Since this is really only useful when you also have --debug turned on, I might require both that Naomi is running in debug mode and that a switch has been set to keep temporary files.

redragonx replied:
Fair enough. I like that plan!

redragonx reviewed (commented):
Everything looks good, ready to approve.

aaronchantrill committed:
DRYing functions and fix local_mic

Changed "raw_input" variable to "text_input" in application.py
to reduce the chance of confusion with the raw_input function
from python2.

Removed "check_option_in_profile" and "get_profile_value"
functions from naomi/diagnose.py and moved them to naomi/profile.py

Renamed "input" variable to "text_input" and renamed "raw_input"
function to "input" in naomi/local_mic.py

Removed get_profile_var and set_profile_var functions from
naomi/populate.py and moved them into naomi/profile.py

Removed diagnose and imported naomi.profile and naomi.testutils
into plugins/stt/pocketsphinx-stt/tests/test_g2p.py

redragonx reviewed (approved):
I like the new changes.
